igvc_navigation/src/field_d_planner/main.cpp

```diff
@@ -0,0 +1,266 @@
+/**
+Solves for an optimal path using the Field D* incremental search algorithm.
+
+Field D* implementation details can be found in FieldDPlanner.h
+
+Author: Alejandro Escontrela <aescontrela3@gatech.edu>
+Date Created: January 23rd, 2018
+*/
+
+#include <geometry_msgs/PoseStamped.h>
+#include <igvc_msgs/map.h>
+#include <nav_msgs/Path.h>
+#include <sensor_msgs/Image.h>
+#include <std_msgs/Int32.h>
+
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
+#include <ros/ros.h>
+#include <tf/transform_datatypes.h>
+#include <igvc_utils/NodeUtils.hpp>
+
+#include <limits>
+#include <mutex>
+#include <tuple>
+#include <vector>
+#include "FieldDPlanner.h"
+#include "Graph.h"
+
+std::mutex planning_mutex;
+
+igvc_msgs::mapConstPtr map;  // Most up-to-date map
+FieldDPlanner planner;       // Field D* path planner
+int x_initial, y_initial;    // Index for initial x and y location in search space
+
+double maximum_distance;  // maximum distance to goal node before warning messages spit out
+
+bool initialize_graph = true;   // set to true if the graph must be initialized
+bool initial_goal_set = false;  // true if the first goal has been set
+bool goal_changed = false;      // true if the goal changed and the graph must be re-initialized
+
+//-------------------------- Helper Methods ----------------------------//
+
+/**
+Publish expanded nodes for visualization purposes. This is not a subscriber
+callback.
+
+@param[in] inds the indices of Nodes that have been expanded in the graph search
+@param[in] expanded_cloud the PCL pointcloud which expanded node indices
+        should be stored in
+@param[in] the publishes with which to publish the PCL pointcloud of expanded nodes
+*/
+void publish_expanded_set(const std::vector<std::tuple<int, int>>& inds,
+                          pcl::PointCloud<pcl::PointXYZRGB>& expanded_cloud, ros::Publisher& expanded_pub)
+{
+  expanded_cloud.clear();
+  expanded_cloud.header.frame_id = "odom";
+
+  for (std::tuple<int, int> ind : inds)
+  {
+    pcl::PointXYZRGB p;
+
+    p.x = static_cast<float>(std::get<0>(ind) - x_initial) * map->resolution;
+    p.y = static_cast<float>(std::get<1>(ind) - y_initial) * map->resolution;
+    p.z = -0.05f;
```

Comment:
Is this arbitrary?

Reply:
Yeah the `p.z = -0.05f` is arbitrary. I didn't want the visualization of expanded nodes to conflict with the path visualization so I just put the expanded pointcloud below.